servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -412,20 +445,74 @@ public final ServerContext listenBlockingStreamingAndAwait(
+
+        if (nonOffloadingFilters.isEmpty()) {
+            filteredService = serviceFilters.isEmpty() ? rawService : buildService(serviceFilters, rawService);
+        } else {
+            boolean anyOffloads = strategy.isSendOffloaded() ||
+                    strategy.isMetadataReceiveOffloaded() ||
+                    strategy.isDataReceiveOffloaded();
+
+            if (anyOffloads) {
+                // We are going to have to offload, even if just to the raw service
+                nonOffloadingFilters.add(new OffloadingFilter(strategy,
+                        buildFactory(serviceFilters)));
+                strategy = null != strategy.executor() ?
+                        HttpExecutionStrategies.customStrategyBuilder()
+                                .offloadNone().executor(strategy.executor()).build() :
+                        HttpExecutionStrategies.noOffloadsStrategy();
+            } else {
+                // All the filters can be appended.
+                nonOffloadingFilters.addAll(serviceFilters);
+            }
+            filteredService = buildService(nonOffloadingFilters, rawService);
         }
-        StreamingHttpService filteredService = currServiceFilter != null ?
-                currServiceFilter.create(rawService) : rawService;
+
         return doListen(connectionAcceptor, filteredService, strategy, drainRequestPayloadBody);
     }
 
-    private static StreamingHttpServiceFilterFactory appendFilter(
-            @Nullable final StreamingHttpServiceFilterFactory current,
-            final StreamingHttpServiceFilterFactory next) {
-        return current == null ? next : service -> current.create(next.create(service));
+    private HttpExecutionStrategyInfluencer getStrategyInfluencer(Object service) {
+        HttpExecutionStrategyInfluencer influencer =
+                buildInfluencer(serviceFilters, strategy -> influenceStrategy(service, strategy));
+        HttpExecutionStrategy useStrategy = influencer.influenceStrategy(strategy);
+
+        return s -> s.merge(useStrategy);
+    }
+
+    private static StreamingHttpServiceFilterFactory buildFactory(Deque<StreamingHttpServiceFilterFactory> filters) {
+        return filters.stream()
+                .reduce((prev, filter) -> strategy -> prev.create(filter.create(strategy)))
+                .orElse(StreamingHttpServiceFilter::new); // unfortunate that we need extra layer
+    }
+
+    private static StreamingHttpService buildService(Deque<StreamingHttpServiceFilterFactory> filters,
+                                                     StreamingHttpService service) {
+        return filters.stream()
+                .reduce((prev, filter) -> svc -> prev.create(filter.create(svc)))
+                .map(factory -> (StreamingHttpService) factory.create(service))
+                .orElse(service);
+    }
+
+    private static HttpExecutionStrategyInfluencer buildInfluencer(Deque<StreamingHttpServiceFilterFactory> filters,
```

Comment:
Originally it was replaced for the optimization of moving filters from offloading to non-offloading and since the data structure was redundant with the filter list  Beyond that it was redundant with the filter list. At the moment the removal is an unnecessary simplification.

Reply:
If there is no change in behavior, would be great to have less duplicated logic between `buildInfluencer` and `StrategyInfluencerChainBuilder`

Reply:
Agreed with @idelpivnitskiy. We now have `StrategyInfluencerChainBuilder` and this `ArrayDeque` approach. Can you confirm the end sate will involve a single way this is accomplished?

Reply:
`StrategyInfluencerChainBuilder` is also used by the client builder and won't be refactored in the current PR. My preference is still for the approach now used in `HttpServerBuilder` as the implementation is simpler and the "rules" are better consolidated to two methods vs divided between `HttpServerBuilder` and `StrategyInfluencerChainBuilder` implementation. The "rules" are:
1. If a service filter implements `HttpStrategyInfluencer` apply the influence in result otherwise apply `defaultStreamingInfluencer()` influence
2. The service is prepended as an influencer if it implements `HttpStrategyInfluencer` otherwise it has no influence
3. The influencer used for non-async streaming adaptor is the combination of the filters' StrategyInfluence (1) and the service's influence (2) applied to the builder strategy. 
4. The builder influencer (3) is used to influence the strategy used by the adaptor using `HttpExecutionStrategy.merge()`

The `StrategyInfluencerChainBuilder` spreads these rules out across `HttpServerBuilder` append methods, listen methods the final `build` method which makes it hard to be sure the behavior is even consistent. The implementation could be simplified a bit, calculating the influence directly rather than building the influencer and then calculating but that didn't seem worth the extra time.

Reply:
not saying the duplication needs to be addressed in this PR, and no objection to the new implementation ... just want to make sure we will eventually have one implementation. since we are introducing duplication in this PR consider filing an issue associated with the current offloading API scope for followup.

Reply:
This will likely need to be addressed in #1695 with the changes to `HttpStrategyInfluencer`